Transitive closure in where clauses

pontusmelke opened the pull request:
For queries like

```
MATCH (n)-->(m) WHERE n.prop = m.prop AND m.prop = 42 RETURN *
```

we can safely rewrite the query to

```
MATCH (n)-->(m) WHERE n.prop = 42 AND m.prop = 42 RETURN *
```

Also rewrites queries of the form

```
MATCH (n)-->(m) WHERE m.prop = n.prop AND m.prop = 42 RETURN *
```

which is turned into

```
MATCH (n)-->(m) WHERE m.prop = 42 AND n.prop = 42 RETURN *
```

pontusmelke committed:
Transitive closure in where clauses

For queries like

```
MATCH (n)-->(m) WHERE n.prop = m.prop AND m.prop = 42 RETURN *
```

we can safely rewrite the query to

```
MATCH (n)-->(m) WHERE n.prop = 42 AND m.prop = 42 RETURN *
```

craigtaverner commented on community/cypher/cypher-compiler-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/ast/rewriters/TransitiveClosureTest.scala:
```diff
@@ -0,0 +1,84 @@
+
+  test("MATCH (a)-->(b) WHERE b.prop = a.prop AND b.prop = 42") {
+    shouldRewrite(
+      "MATCH (a)-->(b) WHERE a.prop = b.prop AND b.prop = 42",
```
Test title differs in equality order, but actual test content does not and is identical to previous test.

pontusmelke replied:
You might almost think I just copy-pasted without thinking about it...

craigtaverner reviewed (approved):
Cool solution. One test might like a small change.

pontusmelke committed:
Handle emerging equalities as well

Add rewriting to also handle the case whera `B == A AND B == 42`
which is turned into `B == 42 AND A == 42`

systay commented on community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/rewriters/transitiveClosure.scala:
```diff
@@ -0,0 +1,105 @@
+    })
+
+    @tailrec
+    private def recurse(in: Where, f: Where => Where): Where = {
```
Isn't this what `fixedPoint` already does?

pontusmelke replied:
Probably I was looking for something called `iterateUntilConvergence` but it was probably renamed years ago.

systay commented on community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/rewriters/transitiveClosure.scala:
```diff
@@ -0,0 +1,105 @@
+
+    def apply(that: AnyRef): AnyRef = instance.apply(that)
+
+    private val instance: Rewriter = topDown(Rewriter.lift {
```
Is there a reason we do this `topDown` instead of `bottomUp`? I was under the impression that `bottomUp` is preferable from a performance perspective, if either work.

systay commented on community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/rewriters/transitiveClosure.scala:
```diff
@@ -0,0 +1,105 @@
+
+    private val instance: Rewriter = topDown(Rewriter.lift {
+      case where@Where(expression) =>
+        recurse(where, w => w.endoRewrite(whereRewriter))
```
For other AST rewriters, we do the fixedPoint recursion on the outside, so rewriters can help each other. We could do that by moving this into the lateAstRewriting phase. Is that possible?

pontusmelke replied:
It is probably possible, however I wanted to do the rewrite before we rewrite `Equals` to `In`.

systay replied:
Ah, makes sense. :+1:

systay reviewed (changes requested):
Really cool! Just a few tweaks.

systay commented:
I would also like to point out that this does not solve another class of transitive closures. This query:
```
MATCH (a), (b) WHERE a = b RETURN a, b
```

is currently solved with this plan:

```
+-----------------+----------------+-----------+-------+
| Operator        | Estimated Rows | Variables | Other |
+-----------------+----------------+-----------+-------+
| +ProduceResults |              0 | a, b      |       |
| |               +----------------+-----------+-------+
| +ValueHashJoin  |              0 | a -- b    | a = b |
| |\              +----------------+-----------+-------+
| | +AllNodesScan |              1 | b         |       |
| |               +----------------+-----------+-------+
| +AllNodesScan   |              1 | a         |       |
+-----------------+----------------+-----------+-------+
```

We could be plan this better than that. This PR is a great step forward. Just wanted to remind us that there is more transitive stuff we can do.

pontusmelke committed:
Fixes from code review

pontusmelke commented:
@systay Agree, handling th

pontusmelke commented:
@systay I agree but that is not really transitive closure. We should probably just rewrite equality of nodes or relationships to  projections, i.e.
`a = b => WITH b AS a`?
